test/puppetlabs/puppetdb/testutils/services.clj

```diff
@@ -133,39 +134,146 @@
 
+(defn create-url-str [base-url url-suffix]
+  (str (utils/base-url->str base-url)
+       (when url-suffix
+         url-suffix)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;
+;; Functions that return URLs and strings for the top level services
+;; in PDB
+;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
 (defn pdb-query-url []
   (assoc *base-url* :prefix "/pdb/query" :version :v4))
 
+(defn query-url-str [url-suffix]
+  (create-url-str (pdb-query-url) url-suffix))
+
 (defn pdb-cmd-url []
   (assoc *base-url* :prefix "/pdb/cmd" :version :v1))
 
+(defn cmd-url-str [url-suffix]
+  (create-url-str (pdb-cmd-url) url-suffix))
+
 (defn pdb-admin-url []
   (assoc *base-url* :prefix "/pdb/admin" :version :v1))
 
-(defn get-url [base-url suffix & [opts]]
-  (let [opts (or opts {:throw-exceptions true
-                       :throw-entire-message? true})]
-    (-> (str (base-url->str base-url) suffix)
-        (client/get opts)
+(defn admin-url-str [url-suffix]
+  (create-url-str (pdb-admin-url) url-suffix))
+
+(defn pdb-metrics-url []
+  (assoc *base-url* :prefix "/metrics" :version :v1))
+
+(defn metrics-url-str [url-suffix]
+  (create-url-str (pdb-metrics-url) url-suffix))
+
+(defn pdb-meta-url []
+  (assoc *base-url* :prefix "/pdb/meta" :version :v1))
+
+(defn meta-url-str [url-suffix]
+  (create-url-str (pdb-meta-url) url-suffix))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;
+;; Basic HTTP functions for interacting with PuppetDB services, the
+;; url-str arguments below will likely come from the above functions
+;; relating to the various top level services
+;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(pls/defn-validated get-unparsed
+  "Executes a GET HTTP request against `url-str`. `opts` are merged
+  into the request map. The response is unparsed but returned as a
+  string."
+  [url-str :- String
+   & [opts]]
+  (http/get url-str
+            (merge
+             {:as :text
+              :headers {"Content-Type" "application/json"}}
+             opts)))
+
+(pls/defn-validated get
+  "Executes a GET HTTP request against `url-str`. `opts` are merged
+  into the request map. JSON responses are automatically parsed before
+  returning. Error responses are returned (i.e. 404) and not thrown."
+  [url-str :- String
+   & [opts]]
+  (let [resp (get-unparsed url-str opts)]
+    (if (testutils/json-content-type? resp)
+      (update resp :body #(json/parse-string % true))
+      resp)))
+
+(def default-ca-cert "test-resources/puppetserver/ssl/certs/ca.pem")
+(def default-cert "test-resources/puppetserver/ssl/certs/localhost.pem")
+(def default-ssl-key "test-resources/puppetserver/ssl/private_keys/localhost.pem")
+
+(pls/defn-validated get-ssl
+  "Executes a mutually authenticated GET HTTPS request against
+  `url-str` using the above `get` function."
+  [url-str & [opts]]
+  (get-ssl url-str
```

Comment:
D'oh, not what I intended. Bummer I can't test this without your branch. I'll fix this up